File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/lc-all-solutions/141.linked-list-cycle/linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

c_ Solution(o..
  ___ hasCycle  head
    """
    :type head: ListNode
    :rtype: bool
    """
    fast slow head
    w.... fast a.. fast.next:
      fast fast.next.next
      slow slow.next
      __ slow __ fast:
        r.. T..
    r.. F..

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/Array/LinkedListCycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */

// 两个指针，一个一次走一步，一个一次走两步，如果有环会在某处相遇。
 var hasCycle = function(head) {
    __ (!head) {
        r_ false
    }
    let oneIndex = head
    let twoIndex = head.next

    __ (!twoIndex) {
        r_ false
    }

    _____ (oneIndex && twoIndex) {
        __ (oneIndex === twoIndex) {
            r_ true
        }

        oneIndex = oneIndex.next
        twoIndex = twoIndex.next

        __ (!twoIndex || !twoIndex.next) {
            r_ false
        }

        twoIndex = twoIndex.next
    }

    r_ false
};